packages/next/lib/recursive-delete.ts

```diff
@@ -0,0 +1,38 @@
+import fs from 'fs'
+import { join } from 'path'
+import { promisify } from 'util'
+
+const readdir = promisify(fs.readdir)
+const stat = promisify(fs.stat)
+const rmdir = promisify(fs.rmdir)
+const unlink = promisify(fs.unlink)
+
+/**
+ * Recursively delete directory contents
+ * @param  {string} dir Directory to delete the contents of
+ * @param  {RegExp} filter Filter for the file name, only the name part is considered, not the full path
+ * @returns Promise void
+ */
+export async function recursiveDelete(dir: string, filter?: RegExp, ensure?: boolean): Promise<void> {
+  let result
+  try {
+    result = await readdir(dir)
+  } catch (e) {
+    if (ensure) throw e
+    // If the dir does not exist we dont need to do anything
+    return
+  }
+
+  await Promise.all(result.map(async (part: string) => {
+    const absolutePath = join(dir, part)
+    const pathStat = await stat(absolutePath)
+
+    if (pathStat.isDirectory()) {
+      return recursiveDelete(absolutePath, filter).then(async () => await rmdir(absolutePath))
```

Comment:
This wouldn't be faster than doing
```js
await recursiveDelete(absolutePath, filter);
return rmdir(absolutePath);
```
right? I mean, the returned promise would take just as long to resolve.

Reply:
True but I was also emphasizing the fact that its removing *after* recursively deleting. IIRC typescript didn't like returning `rmdir` because it has the wrong return type